Sources/XcodeProjToGraph/Mappers/Build/BuildPhaseMapper.swift

```diff
@@ -0,0 +1,428 @@
+    /// - Returns: An array of `CopyFilesAction` instances describing how files are copied at build time.
+    /// - Throws: If file references or paths cannot be resolved.
+    func mapCopyFiles(target: PBXTarget) async throws -> [CopyFilesAction]
+
+    /// Maps Core Data models from the target’s resource phases.
+    ///
+    /// - Parameter target: The Xcode target to map.
+    /// - Returns: An array of `CoreDataModel` instances describing the target’s Core Data models.
+    /// - Throws: If model paths cannot be resolved.
+    func mapCoreDataModels(target: PBXTarget) async throws -> [CoreDataModel]
+
+    /// Maps raw script build phases for debugging or analysis.
+    ///
+    /// - Parameter target: The Xcode target to map.
+    /// - Returns: An array of `RawScriptBuildPhase` instances representing the raw script phases.
+    func mapRawScriptBuildPhases(target: PBXTarget) async throws -> [RawScriptBuildPhase]
+
+    /// Maps additional files that are not included in any build phase.
+    ///
+    /// - Parameter target: The Xcode target to map.
+    /// - Returns: An array of `FileElement` for files that are part of the project but not in build phases.
+    /// - Throws: If file references or paths cannot be resolved.
+    func mapAdditionalFiles(target: PBXTarget) async throws -> [FileElement]
+
+    /// Maps frameworks referenced by the target’s frameworks build phase.
+    ///
+    /// - Parameter target: The Xcode target to map.
+    /// - Returns: An array of `TargetDependency` instances representing frameworks.
+    /// - Throws: If framework references or paths cannot be resolved.
+    func mapFrameworks(target: PBXTarget) async throws -> [TargetDependency]
+}
+
+/// A mapper responsible for converting the build phases of a `PBXTarget` into corresponding domain models.
+///
+/// The `BuildPhaseMapper` leverages `ProjectProviding` to resolve file paths and uses XcodeProj’s APIs to
+/// navigate build phases (sources, resources, headers, scripts, copy files, and frameworks).
+///
+/// By producing strongly-typed domain models (`SourceFile`, `ResourceFileElement`, `Headers`, `TargetScript`,
+/// etc.), it enables subsequent steps (like code generation, analysis, or custom tooling) to operate on a well-structured,
+/// semantic representation of the target’s build phases.
+public final class BuildPhaseMapper: BuildPhaseMapping {
+    private let projectProvider: ProjectProviding
+
+    /// Creates a new `BuildPhaseMapper`.
+    ///
+    /// - Parameter projectProvider: Provides access to the project’s paths, files, and parsed structures.
+    public init(projectProvider: ProjectProviding) {
+        self.projectProvider = projectProvider
+    }
+
+    public func mapSources(target: PBXTarget) async throws -> [SourceFile] {
+        guard let sourcesPhase = try target.sourcesBuildPhase() else { return [] }
+        return try await sourcesPhase.files?.asyncCompactMap { try await self.mapSourceFile($0) }
```

Comment:
💯 I ran some comparisons and found that running the Tuist repo's Xcode workspace with the old async implementation took about 2 seconds, while the sync version took around 5 seconds. However, for smaller projects, the overhead of async/await made the sync version noticeably faster. Given this, I’ve removed the async code for now, as it does feel like it might have been a case of premature optimization

Reply:
Let's revisit this later on – those numbers are not nothing, so I think we should consider some parallelization. But let's first get a stable synchronous version going and then we can look into parallelizing the parts where it would make a big difference.